experimental/dds/tree2/src/core/undo/undoRedoManager.ts

```diff
@@ -68,27 +71,117 @@ export class UndoRedoManager<TChange, TEditor extends ChangeFamilyEditor> {
 
 		const { commit, repairData } = commitToUndo;
 
-		return this.changeFamily.rebaser.invert(
+		let change = this.changeFamily.rebaser.invert(
 			tagChange(commit.change, commit.revision),
 			false,
 			repairData,
 		);
+
+		// Rebase the inverted change onto any commits that occurred after the undoable commits.
+		const head = this.getHead();
+		if (head !== undefined && commit.revision !== head.revision) {
+			const pathAfterUndoable: GraphCommit<TChange>[] = [];
+			const ancestor = findCommonAncestor([commit], [head, pathAfterUndoable]);
+			assert(ancestor === commit, "The head commit should be based off the undoable commit.");
+			change = pathAfterUndoable.reduce(
+				(a, b) => this.changeFamily.rebaser.rebase(a, b),
+				change,
+			);
+		}
+
+		return change;
 	}
 
 	/**
 	 * Creates a copy of this `UndoRedoManager`.
-	 * @param repairDataStoreProvider - Optional {@link IRepairDataStoreProvider} to use for the new `UndoRedoManager`.
-	 * If one is not provided, the `repairDataStoreProvider` of this `UndoRedoManager` will be cloned.
+	 * @param getHead - Optional function to use for retrieving the head commit for the branch associated with the 
+	 * the new {@link UndoRedoManager}. If one is not provided, the one from this {@link UndoRedoManager} will be used.
+	 * @param repairDataStoreProvider - Optional {@link IRepairDataStoreProvider} to use for the new {@link UndoRedoManager}.
+	 * If one is not provided, the `repairDataStoreProvider` of this {@link UndoRedoManager} will be cloned.
+	 * @param headUndoableCommit - Optional head undoable commit, if one is not provided the head undoable commit
+	 * of this {@link UndoRedoManager} will be used.
 	 */
 	public clone(
+		getHead?: () => GraphCommit<TChange> | undefined,
 		repairDataStoreProvider?: IRepairDataStoreProvider,
+		headUndoableCommit?: UndoableCommit<TChange>,
 	): UndoRedoManager<TChange, TEditor> {
 		return new UndoRedoManager(
 			repairDataStoreProvider ?? this.repairDataStoreProvider.clone(),
 			this.changeFamily,
-			this.headUndoableCommit,
+			getHead ?? this.getHead,
+			headUndoableCommit ?? this.headUndoableCommit,
 		);
 	}
+
+	/**
+	 * Updates the state of this {@link UndoRedoManager} to correctly reference commits that have been rebased after merging.
+	 * @param baseHead - the head commit of the branch that was rebased onto.
+	 * @param rebasedHead - the head commit of the newly rebased branch.
+	 * @param originalUndoRedoManager - the {@link UndoRedoManager} of the branch that was rebased.
+	 */
+	public updateAfterMerge(
+		baseHead: GraphCommit<TChange>,
+		rebasedHead: GraphCommit<TChange>,
```

Comment:
I'm getting a little confused by these names. What do you think about `rebasedHead` -> `newHead` and `originalUndoRedoManager` -> `mergedUndoRedoManager`?